### ISSUE TITLE
Where to use navigation methods? (Question)

### ISSUE THREAD

**Issue by stefandevo** [jonataslaw/get#140]
Navigating to screens, would you do this in the Controller (as it is “business logic”)? 
However, when not using named routes, it would mean I have to use the Widget (Page Widget) in a Controller. What do you prefer? Navigating inside the Widgets, or in the Controllers?

**Comment by jonataslaw** [comment]
Well, navigation is not business logic, it is navigation logic.
There is no explicit prohibition in any convention, from the use of navigation logic in View, or in Controllers, because it is not part of either. It can be either in View, or in Controller, or separate from both.
If you use onPressed to navigate, that's OK, the problem arises when navigation depends on business logic, and this makes navigation in View not a good idea.
I have seen examples of users of this lib, as well as others, where the navigation logic is inside an AnyBuilder(builder: x).
I don't think it's cool, because it can become a memoryLeak factory. Many people simply use this, and don't mind checking the Dart inspector and seeing that nothing inside a builder method is laid out, as I explained in this comment https://github.com/jonataslaw/get/issues/136# issuecomment-633624493

Obviously, if it is a variable, it will not make much difference, the problem is in inserting entire stacks within Builders methods, such as:
```dart
if (UserStatus.logged) {
return Home ();
} else {
return Login ();
}
```

The first time you access this, the compiler will only go through the Login();
If the authenticated state changes, it will receive a setState, and the compiler will go to the Home() line, however, the login stack will remain whole in memory, because setState is not able to dispose widgets.
If the user logs in and out using each other, the app will double the memory spent on each process.
Obviously, this could be remedied using a Get.off/pushReplacement instead of returning a widget, but we went into the direct infringement of the Flutter documentation that says builder methods should only receive widgets, and using Navigator 2.0 this is ok, because it is a Widget, using normal navigation, this can be considered a bad practice.
This is a structural problem, and unfortunately it is widely used, even as an "example" in repositories of some state managers, and as I usually say, inserting objects within Builders methods is to manufacture memoryLeaks. In many cases, the GC simply cannot clean this up (a very extreme example of where this could happen, would be to insert a StreamController into it, you will notice that a sink has been opened, and after a setState a new sink has been created, and the old went to a limbo that the GC cannot find it).

I recommend using the navigation logic inside the Controller, or creating a page just for navigation (I will give examples of this as soon as I document the bindings API), but not using it inside a widget. First, because this is a behavior that must be adopted for the entire project (another dev that reads your code will be confused if you insert a navigation in the View, another in the Controller, another in the Repository, etc.), and as shown above, adopt everything in the View can cause problems in some cases, so it would be preferable to call new routes in the Controller, even from an onPressed, it costs nothing to insert the method in your controller, and call from the View, if you don't have a class for that. Another even clear approach is to create a page just for navigation, with all Get.dialog(), Get.to(), Get.anything, and preferably call these actions from the Controller.
But it depends on what structure and architecture you are using, definitely.

**Comment by MichaelOttink** [comment]
It get's kinda difficult with firebase tho. Since it has a _onAuthChanged() which can fire at any moment and route you to a page. So when you put() a authentication controller it in your app init and then build the page it will fire during build a Get.off() to a page.

 So I made it lazy put so that it is called in the splash page for the first time but even then it it is messy. I call it with a builder and that builder used to have the TheHomePage() and the SignUpPage() this will result in the page being built and then being built again instantly. So I added a loading indicator in the GetBuilder so now the builder's only function is to trigger the AuthenticationController. So I change the builder to a Get.find() just to trigger the AuthenticationController. But this get's really messy.

It get's really messy when trying to delete a account. Because you basically call the FirebaseUser.delete() method and that will trigger the onAuthChanged() which will do a Get.off(). I'm not really sure how and when the Get.off() will update but it routes to TheSignUpPage and then throws an error for all the previous widgets because the controllers didn't get removed so all the streams are still intact.
